sycl/include/syclcompat/math.hpp

```diff
@@ -744,15 +796,21 @@ inline std::common_type_t<ValueT, ValueU> fmax_nan(const ValueT a,
                                                    const ValueU b) {
   if (detail::isnan(a) || detail::isnan(b))
     return NAN;
-  return sycl::fmax(static_cast<std::common_type_t<ValueT, ValueU>>(a),
-                    static_cast<std::common_type_t<ValueT, ValueU>>(b));
+  return syclcompat::max(a, b);
 }
+
 template <typename ValueT, typename ValueU>
 inline sycl::vec<std::common_type_t<ValueT, ValueU>, 2>
 fmax_nan(const sycl::vec<ValueT, 2> a, const sycl::vec<ValueU, 2> b) {
   return {fmax_nan(a[0], b[0]), fmax_nan(a[1], b[1])};
 }
 
+template <typename ValueT, typename ValueU>
+inline sycl::marray<std::common_type_t<ValueT, ValueU>, 2>
+fmax_nan(const sycl::marray<ValueT, 2> a, const sycl::marray<ValueU, 2> b) {
+  return {fmax_nan(a[0], b[0]), fmax_nan(a[1], b[1])};
+}
```

Comment:
I liked this idea and I gave it a shot, but unfortunately it's actually quite tricky to implement. In the `unordered_compare_mask` case, we can treat `ValueT` as a concrete class (`sycl::vec<int, 2>`, `sycl::marray<int, 2>`), but in the `fmax_nan` case, we need to be able to reason about the contained types (i.e. `ValueT = int` and `ValueU = short`) because we need to declare the return type to be e.g. `sycl::vec<std::common_type_t<ValueT, ValueU>>`.

So we could have something like this:
```cpp
template <template <typename, int> typename ContainerT, typename ValueT,
          typename ValueU>
inline ContainerT<std::common_type_t<ValueT, ValueU>, 2>
fmax_nan(const ContainerT<ValueT, 2> a, const ContainerT<ValueU, 3> b);
```
but because `sycl::vec` takes `int` and `sycl::marray` takes `size_t`, there isn't a definition of `ContainerT` which catches both.

:melting_face: